### PR TITLE
Add VRFCoordinator and VRFSubscription support

### PR DESCRIPTION
**Pull request by codingSince9**

### CODE REVIEW

File: src/chainlink/VRFCoordinator.ts [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+            ["uint96", "uint64", "address", "address[]"],
+            "0x0"
+        );
+        subscription.balance = callRequest.return_values[0];
+        subscription.requestCount = callRequest.return_values[1];
+        subscription.owner = callRequest.return_values[2];
+        subscription.consumers = Array.from(callRequest.return_values[3]);
+        subscription.coordinatorAddress = this.address;
+        subscription.chainlinkTokenContractAddress = this.chainlinkTokenContractAddress;
+        this.subscriptions.push(subscription);
```
Comment (fichiokaku) [inline src/chainlink/VRFCoordinator.ts:85]: It's cool that you're caching subscriptions per subscription ID and then reusing them when getSubscription is called for the same subscription ID. But I'm afraid that if the subscription data changes, we will still hold the old data using this method. For example, if the subscription balance is reduced after we fetch the object, we will still show the old data - as if the balance was greater than it actually is.

Why don't we remove the caching and the subscriptions array and simply fetch the subscription object from the chain every time the function is called?

Additionally, add the `getInfo()` function on the VRFSubscription object to re-read the values again if we want to see if something changed. This function will call the `readContract("getSubscription")` on the VRFCoordinator again and update its state with the latest subscription data, while also returning the same data to the caller.

File: src/core/types.ts [hunk 2]
```diff
@@ -353,3 +353,20 @@ export interface AssetSendRequest {
...
+  custom_rpc_url?: string;
+  created_at: Date;
+}
+
+export interface VRFCoordinatorConfig {
+  minimumRequestConfirmations: number;
+  maxGasLimit: number;
+  stalenessSeconds: number;
+  gasAfterPaymentCalculation: number;
+}
```
Comment (fichiokaku) [inline src/core/types.ts:372]: add newline

File: src/core/helpers/util.ts [hunk 3]
```diff
@@ -32,3 +34,34 @@ export function ensureBrowser(): void {
...
+  function_params: EncodedFunctionParameter[],
+  eth_amount: string,
+) {
+  return await MainApi.instance().createFunctionCallRequest({
+    contract_address,
+    function_name,
+    function_params,
+    eth_amount,
+  });
+}
```
Comment (fichiokaku) [inline src/core/helpers/util.ts:67]: add newline

File: src/chainlink/VRFSubscription.ts [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+            this.coordinatorAddress,
+            "requestSubscriptionOwnerTransfer",
+            [
+                { type: "uint64", value: this.id },
+                { type: "address", value: newOwner }
+            ],
+            "0"
+        );
+    }
+}
```
Comment (fichiokaku) [inline src/chainlink/VRFSubscription.ts:106]: add newline

File: src/chainlink/VRFSubscription.ts [hunk 5]
```diff
@@ -0,0 +1,106 @@
+import { writeContract } from "../core/helpers/util";
+
+
+export class VRFSubscription {
+    public id: string = "";
+    public balance: string = "";
+    public requestCount: string = "";
+    public owner: string = "";
+    public consumers: string[] = [];
```
Comment (fichiokaku) [inline src/chainlink/VRFSubscription.ts:9]: How is consumers array populated?

File: src/chainlink/VRFSubscription.ts [hunk 6]
```diff
@@ -0,0 +1,106 @@
+import { writeContract } from "../core/helpers/util";
+
+
+export class VRFSubscription {
```
Comment (fichiokaku) [inline src/chainlink/VRFSubscription.ts:4]: move VRFSubscription & VRFCoordinator to a "vrf" subfolder: 
`./chainlink/vrf/VRFCoordinator.ts`
`./chainlink/vrf/VRFSubscription.ts`

File: src/chainlink/VRFCoordinator.ts [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+        const callRequest = await readContract(
+            this.address,
+            "getTotalBalance",
+            [],
+            ["uint96"],
+            "0x0"
+        );
+        return callRequest.return_values[0];
+    }
+}
```
Comment (fichiokaku) [inline src/chainlink/VRFCoordinator.ts:115]: add newline

File: src/chainlink/VRFCoordinator.ts [hunk 8]
```diff
@@ -0,0 +1,115 @@
+import { VRFSubscription } from "./VRFSubscription";
+import { ContractCallAction } from '../core/actions/ContractCallAction';
+import { readContract, writeContract } from '../core/helpers/util';
+import { ethers } from 'ethers';
+import { MainApi } from '../core/api/main-api';
+import { VRFCoordinatorConfig } from '../core/types';
+
+
```
Comment (fichiokaku) [inline src/chainlink/VRFCoordinator.ts:8]: Remove one newline (1 is enough)

File: src/chainlink/VRFCoordinator.ts [hunk 9]
```diff
@@ -0,0 +1,115 @@
...
+let chainlist: any;
+let tokenAndCoordinatorAddresses: any;
+
```
Comment (fichiokaku) [inline src/chainlink/VRFCoordinator.ts:11]: Remove one newline (1 is enough)

File: src/chainlink/VRFCoordinator.ts [hunk 10]
```diff
@@ -0,0 +1,115 @@
...
+}
+
+async function fetchTokenAndCoordinatorAddresses() {
+    const tokenAndCoordinatorAddressesResponse = await fetch(
+        'https://raw.githubusercontent.com/0xpolyflow/polyflow-sdk/master/resources/chainlink_contract_and_coordinator.json'
+    );
+    const tokenAndCoordinatorAddressesJson = await tokenAndCoordinatorAddressesResponse.json();
+    tokenAndCoordinatorAddresses = new Map(Object.entries(tokenAndCoordinatorAddressesJson));
+}
+
```
Comment (fichiokaku) [inline src/chainlink/VRFCoordinator.ts:28]: Remove one newline (1 is enough)

File: src/core/helpers/util.ts [hunk 11]
```diff
@@ -32,3 +34,34 @@ export function ensureBrowser(): void {
...
+) {
+  return await MainApi.instance().readContract({
+    contract_address,
+    function_name,
+    function_params,
+    output_params,
+    caller_address,
+  });
+}
+
```
Comment (fichiokaku) [inline src/core/helpers/util.ts:53]: remove 1 newline (1 is enough)

File: src/core/helpers/util.ts [hunk 12]
```diff
@@ -32,3 +34,34 @@ export function ensureBrowser(): void {
...
+
+export async function writeContract(
+  contract_address: string,
+  function_name: string,
+  function_params: EncodedFunctionParameter[],
+  eth_amount: string,
+) {
+  return await MainApi.instance().createFunctionCallRequest({
```
Comment (fichiokaku) [inline src/core/helpers/util.ts:61]: Wrap the FunctionCallRequest object (returned from MainAPI) inside the ContractCall object, which already contains the present() function used to show the widget.

```
return new ContractCallAction(
    await MainApi.instance().createFunctionCallRequest({
        ...
    })
);
```
Also add the return type of the outer `writeContract` to be set to `Promise<ContractCallAction>`

File: src/chainlink/VRFSubscription.ts [hunk 13]
```diff
@@ -0,0 +1,106 @@
...
+    public requestCount: string = "";
+    public owner: string = "";
+    public consumers: string[] = [];
+    public coordinatorAddress: string = "";
+    public chainlinkTokenContractAddress: string = "";
+
+    constructor() { }
+
+    public async fund(amount: string) {
+        await writeContract(
```
Comment (fichiokaku) [inline src/chainlink/VRFSubscription.ts:16]: Set the return value of the `fund()` function to be the same as the return value of the `writeContract()` function

```
public async fund(amount: string): Promise<ContractCallAction> {
    return await writeContract(...
}
```

Do the same for all the other state changing functions on VRFCoordinator & VRFSubscription objects.

File: src/chainlink/VRFSubscription.ts [hunk 14]
```diff
@@ -0,0 +1,106 @@
+import { writeContract } from "../core/helpers/util";
+
```
Comment (fichiokaku) [inline src/chainlink/VRFSubscription.ts:2]: remove newline (1 is enough)

**Review by fichiokaku — CHANGES_REQUESTED** [review]

File: src/chainlink/VRFSubscription.ts (hunk 5 above)
Comment (codingSince9) [reply src/chainlink/VRFSubscription.ts:9]: There is a line in getSubscription() method of VRFCoordinator
`subscription.consumers = Array.from(callRequest.return_values[3]);`

**Pull request merged**